### PR TITLE
pss-networktest merge to swarm-network-rewrite

### PR DESCRIPTION
**Pull request by nonsense**

### CODE REVIEW

**Comment by nonsense** [comment]
@zelig @nolash I suggest we merge this, so that we have all outstanding Swarm code in `swarm-network-rewrite`, and from there:
1. Resolve conflicts with `go-ethereum/master`
2. Start cherry-picking parts and commits from `swarm-network-rewrite` and merge gradually to `go-ethereum/master`

Let me know if you have other ideas or something against this.

**Comment by lmars** [comment]
@nonsense I think having one big branch is easier to manage, splitting out individual bits for PRs to master is simpler that way.

I really think we should enable CI on that branch too.

**Comment by nonsense** [comment]
@lmars I agree. If @zelig or @nolash also agrees, then I will merge this PR, and look into enabling Travis CI on `swarm-network-rewrite`.

**Comment by nolash** [comment]
I'm not 100% sure how cherry picking commits will look like in practice, I really don't have experience with big code merges such as this.

It's ok for me as long as we're able to isolate the individual branches of development properly, so we don't have to sift through a lot of irrelevant code in the process of reviewing each part. 

Also, I'd imagine if bits and pieces are merged to master, then most of the time the CI tests will be in a failed state. Are they really then that useful?

**Comment by nolash** [comment]
As for this PR, is there actually stuff left to review, or has everything been reviewed in parts?

**Comment by nonsense** [comment]
@nolash I am not sure what has been reviewed from `pss-networktest`, apart from the smaller ones we merged over the last week or two, and mostly PSS, it does include a few other things:
- p2p/simulation changes
- swarm binary changes
- swarm/api changes
- swarm network hive changes

Once we break down `swarm-network-rewrite`, the upcoming PRs will pass through thorough code review.

**Comment by nonsense** [comment]
I don't think it is feasible to review big change-sets either, this is why I want to bring everything together and will gradually cherry-pick small pieces and bring them to master. We will have a hard time doing that from multiple branches though.

As for the CI tests failures - unfortunately they are not trivial to fix, and all failures are due to go-ethereum at the moment. We would just have to be careful and review Travis logs and make sure we don't break Swarm too. So this is ortogonal to what we're doing.

New code should be branched off from `swarm-network-rewrite` (our god Swarm branch) or `go-ethereum/master` depending on how long it takes to bring `swarm-network-rewrite` to `go-ethereum/master` and how coupled the new changes would be to `swarm-network-rewrite`. This way we can try to avoid making the merge to master harder.

I hope all this makes sense :)

**Comment by nolash** [comment]
I think it does :) I guess @gbalint will be the one keeping all this in line when he starts next week? Should he stick a word in here?

**Comment by gbalint** [comment]
Well, I don't wanna formulate strong opinion before I really understand the details.
My 2 cents generally, don't take it too strongly:
- I prefer small PR-s and frequent merges into a single mainline repo. Big PR-s are impossible to review, they are risky, and parallelly maintained long-living branches can be a nightmare to merge. This means if you have a big project it is important to plan how is it possible to split it into small mergeable parts.
- I prefer to run all or nearly all CI tests on PR-s too, and configure github so it only allows to merge if the tests pass. This helps to keep the mainline clean.
- Mandatory code review is a good idea generally, I would even set up code change management policy on github that it is needed for the merge. However, with people working in different timezones this can be painful.
- I generally dislike cherry picking, because it is very hard to follow what is happening.

**Comment by gbalint** [comment]
To the actual problem here I don't fully understand the situation. If I understand well you rarely merge to `master` , there is a long-living shared project branch `swarm-network-rewrite`, and then there are shorter living branches for smaller developments, which you regularly merge to `swarm-network-rewrite`. I guess `swarm-network-rewrite` can't be merge to master, because it is a big refactor, and it is not functional yet. Am I right?
If yes, then I have some questions:
- On which branches you have CI?
- Do you expect `swarm-network-rewrite` to be fully functional on each commit, or is it just a development branch which can be broken?
- Was there a reason for `pss-networktest` to not be merged earlier to `swarm-network-rewrite`, so this PR won't be this huge?

And I have this feeling that maybe the comment section if this PR is not the best place for this general discussion, let's move it to somewhere else if you agree (gitter?)

**Comment by nonsense** [comment]
@gbalint I agree with all you said. To answer your questions:
1. On the forked `ethersphere/go-ethereum` repo we have no CI. We have to enable it on Travis CI.
2. We have at least two long-lived branches right now - `pss-networktest` and `swarm-network-rewrite`, which I want to merge into one, so that we can gradually merge to master and move to the development flow you described - smaller, more frequent PRs.
3. `swarm-network-rewrite` cannot be merged to master right now for number of reasons:
- it has conflicts with master
- it is too big to review
- AFAIK some functionality is missing from it (syncing)
4. I don't think we expect `swarm-network-rewrite` to be fully functional on each commit, the aim is to enable CI once we resolve merge conflicts with master and move on from there.
5. Reason for `pss-networktest` to not be merged earlier? I don't know. I might have contributed to the problem though, because I worked on failing tests on `pss-networktest` and ultimately merged a few PRs to it, making the change-set even bigger.

---

To summarize the idea for moving forward:
1. Bring all changes to one god branch `swarm-network-rewrite`, and resolve its conflicts from master.
2. Enable CI on this branch (and preferably all other branches in `ethersphere/go-ethereum`)
3. Split `swarm-network-rewrite` into smaller PRs, merge then to `go-ethereum/master` and try to keep future PRs small and frequent towards `go-ethereum/master`.

**Comment by nolash** [comment]
@nonsense you are aware that `pss` was supposed to be the long-lived branch for that working group? But that branch should *always* be functional, so that people doing cutting edge dev on `pss` can rely on it.

**Comment by nolash** [comment]
@nonsense is this now obsoleted by #157 ?

**Comment by gbalint** [comment]
@nolash this is a PR to `swarm-network-rewrite` , while https://github.com/ethersphere/go-ethereum/pull/157 targets `pss`, or do I misunderstand something?

**Comment by nolash** [comment]
@gbalint my bad, you're right

**Comment by nonsense** [comment]
I'm a bit lost on what is going on here. Suddently the number of changes is drastically more than before and we have conflicts. How shall we proceed with this PR ? cc @nolash @gbalint

**Comment by gbalint** [comment]
@nonsense I agree, something is not OK with this PR. I can not believe there are 250+ commit, 653 file diff between just two `pss` branches.

**Comment by gbalint** [comment]
And 85K added lines? Sounds impossible.

**Comment by gbalint** [comment]
Maybe the problem is that `pss` was not synced with `master`? Because the diff is full of commits from master nothing to do with pss.

**Comment by gbalint** [comment]
Ok, so i mixed up this PR with https://github.com/ethersphere/go-ethereum/pull/157
But that one is huge too. First solve the other one to have a single branch for pss related stuff

**Comment by nolash** [comment]
Obsoleted by https://github.com/ethersphere/go-ethereum/pull/168 (which steps through merge with `pss` first)
